components/tidb_query_vec_expr/src/impl_string.rs

```diff
@@ -558,214 +489,187 @@ fn elt_validator(expr: &tipb::Expr) -> Result<()> {
-    if let (Some(arg), Some(pat), Some(direction)) = (arg, pat, direction) {
-        match TrimDirection::from_i64(*direction) {
-            Some(d) => {
-                let arg = String::from_utf8_lossy(arg);
-                let pat = String::from_utf8_lossy(pat);
-                Ok(Some(trim(&arg, &pat, d)))
-            }
-            _ => Err(box_err!("invalid direction value: {}", direction)),
-        }
-    } else {
-        Ok(None)
+    arg: BytesRef,
+    pat: BytesRef,
+    direction: &i64,
+    writer: BytesWriter,
+) -> Result<BytesGuard> {
+    match TrimDirection::from_i64(*direction) {
+        Some(d) => {
+            let arg = String::from_utf8_lossy(arg);
+            let pat = String::from_utf8_lossy(pat);
+            Ok(writer.write(Some(trim(&arg, &pat, d))))
+        }
+        _ => Err(box_err!("invalid direction value: {}", direction)),
     }
 }
 
-#[rpn_fn(nullable)]
+#[rpn_fn]
 #[inline]
-pub fn char_length(bs: Option<BytesRef>) -> Result<Option<Int>> {
-    Ok(bs.map(|b| b.len() as i64))
+pub fn char_length(bs: BytesRef) -> Result<Option<Int>> {
+    Ok(Some(bs.len() as i64))
 }
 
-#[rpn_fn(nullable)]
+#[rpn_fn]
 #[inline]
-pub fn char_length_utf8(bs: Option<BytesRef>) -> Result<Option<Int>> {
-    match bs {
-        Some(bytes) => match str::from_utf8(bytes) {
-            Ok(s) => Ok(Some(s.chars().count() as i64)),
-            Err(err) => Err(box_err!("invalid input value: {:?}", err)),
-        },
-        _ => Ok(None),
+pub fn char_length_utf8(bs: BytesRef) -> Result<Option<Int>> {
+    match str::from_utf8(bs) {
+        Ok(s) => Ok(Some(s.chars().count() as i64)),
+        Err(err) => Err(box_err!("invalid input value: {:?}", err)),
     }
 }
 
-#[rpn_fn(nullable)]
+#[rpn_fn(writer)]
 #[inline]
-pub fn to_base64(bs: Option<BytesRef>) -> Result<Option<Bytes>> {
-    match bs {
-        Some(bytes) => {
-            if bytes.len() > tidb_query_datatype::MAX_BLOB_WIDTH as usize {
-                return Ok(Some(Vec::new()));
-            }
-
-            if let Some(size) = encoded_size(bytes.len()) {
-                let mut buf = vec![0; size];
-                let len_without_wrap =
-                    base64::encode_config_slice(bytes, base64::STANDARD, &mut buf);
-                line_wrap(&mut buf, len_without_wrap);
-                Ok(Some(buf))
-            } else {
-                Ok(Some(Vec::new()))
-            }
-        }
-        None => Ok(None),
+pub fn to_base64(bs: BytesRef, writer: BytesWriter) -> Result<BytesGuard> {
+    if bs.len() > tidb_query_datatype::MAX_BLOB_WIDTH as usize {
+        return Ok(writer.write_ref(Some(b"")));
     }
+
+    let result = if let Some(size) = encoded_size(bs.len()) {
+        let mut buf = vec![0; size];
+        let len_without_wrap = base64::encode_config_slice(bs, base64::STANDARD, &mut buf);
+        line_wrap(&mut buf, len_without_wrap);
+        buf
+    } else {
+        b"".to_vec()
+    };
+
+    Ok(writer.write(Some(result)))
```

Comment:
You may write two return statements respectively in L651 and L653, so as to eliminate the need of allocating a null vector.